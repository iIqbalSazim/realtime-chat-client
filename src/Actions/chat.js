export const changeChatUser = (user) => ({
  type: "CHANGE_CHAT_USER",
  payload: user,
});

export const clearChatUser = () => ({
  type: "CLEAR_CHAT_USER",
});
