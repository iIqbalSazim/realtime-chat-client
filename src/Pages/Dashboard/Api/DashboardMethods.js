import { getRequest, postRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (receiverId) => {
  return await getRequest(`messages?chatting_with=${receiverId}`);
};

export const sendMessage = async (message) => {
  return await postRequest("messages", message);
};
