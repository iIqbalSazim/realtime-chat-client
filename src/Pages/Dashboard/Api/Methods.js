import { getRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (senderId, receiverId) => {
  return await getRequest(
    `messages?sender_id=${senderId}&receiver_id=${receiverId}`
  );
};
