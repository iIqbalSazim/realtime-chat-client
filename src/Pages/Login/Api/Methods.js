import { postRequest } from "../../../Config/Axios/AxiosConfig";

export const loginUser = async (user) => {
  return await postRequest("login", user);
};

export const revokeToken = async (token) => {
  return await postRequest("revoke_token", token);
};
