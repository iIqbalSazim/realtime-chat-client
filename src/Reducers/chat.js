const initialState = {
  user: null,
};

const chatReducer = (state = initialState, action) => {
  switch (action.type) {
    case "CHANGE_CHAT_USER":
      return {
        ...state,
        user: action.payload,
      };
    case "CLEAR_CHAT_USER":
      return {
        ...state,
        user: null,
      };
    default:
      return state;
  }
};

export default chatReducer;
