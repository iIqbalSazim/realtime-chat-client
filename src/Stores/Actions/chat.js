export const changeChatUser = (user) => ({
  type: "CHANGE_CHAT_USER",
  payload: user,
});

export const clearChatUser = () => ({
  type: "CLEAR_CHAT_USER",
});

export const setMessages = (messages) => ({
  type: "SET_MESSAGES",
  payload: messages,
});
