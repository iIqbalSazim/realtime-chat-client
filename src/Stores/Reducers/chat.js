const initialState = {
  messages: [],
  chatUser: null,
};

const chatReducer = (state = initialState, action) => {
  switch (action.type) {
    case "SET_MESSAGES":
      return {
        ...state,
        messages: [...action.payload],
      };
    case "CHANGE_CHAT_USER":
      return {
        ...state,
        chatUser: action.payload,
      };
    case "CLEAR_CHAT_USER":
      return {
        ...state,
        chatUser: null,
      };
    default:
      return state;
  }
};

export default chatReducer;
