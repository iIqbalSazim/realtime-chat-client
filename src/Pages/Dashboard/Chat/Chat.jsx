/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect, useState } from "react";
import {
  ChatContainer,
  MessageList,
  Message,
  ConversationHeader,
  MessageInput,
} from "@chatscope/chat-ui-kit-react";

import { fetchUserMessageHistory } from "../Api/Methods";
import { convertMessageFormat } from "../DashboardHelpers";

const Chat = ({ chattingWith, currentUser }) => {
  const [messageHistory, setMessageHistory] = useState([]);

  const getMessageHistory = async () => {
    const response = await fetchUserMessageHistory(
      currentUser.id,
      chattingWith.id
    );

    const apiMessages = response.data;

    const chatMessages = convertMessageFormat(
      apiMessages,
      currentUser,
      chattingWith
    );

    setMessageHistory(chatMessages);
  };

  useEffect(() => {
    getMessageHistory();
  }, [chattingWith]);

  return (
    <ChatContainer>
      <ConversationHeader className="mb-2">
        <ConversationHeader.Content userName={chattingWith.username} />
      </ConversationHeader>
      <MessageList>
        {messageHistory.length !== 0 ? (
          messageHistory.map((message) => (
            <Message key={message.id} model={{ ...message }} />
          ))
        ) : (
          <MessageList.Content className="items-center flex justify-center w-full my-2">
            <div className="btn btn-active btn-ghost btn-wide">Say Hi!</div>
          </MessageList.Content>
        )}
      </MessageList>
      <MessageInput attachButton={false} placeholder="Type message here" />
    </ChatContainer>
  );
};

export default Chat;
