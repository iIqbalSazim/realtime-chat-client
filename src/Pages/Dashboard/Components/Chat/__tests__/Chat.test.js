import { configure, render, screen, waitFor } from "@testing-library/react";
import { act } from "react-test-renderer";
import ActionCable from "actioncable";

import { renderWithProviders } from "../../../../../../jest.setup";

import { setMessages } from "../../../../../Stores/Actions/chat";
import {
  mockChatMessages,
  mockNewMessage,
} from "../../../../../MockData/messageItems";
import * as ApiMethods from "../../../Api/DashboardMethods";
import {
  mockChattingWith,
  mockCurrentUser,
} from "../../../../../MockData/userItems";
import Chat from "../Chat";
import userEvent from "@testing-library/user-event";

configure({ testIdAttribute: "data-placeholder" });

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn().mockReturnValue(jest.fn()),
  useDispatch: jest.fn().mockReturnValue(jest.fn()),
}));

describe("Chat", () => {
  test("renders correctly with populated messages", async () => {
    jest
      .spyOn(require("react-redux"), "useSelector")
      .mockReturnValue([...mockChatMessages]);
    jest
      .spyOn(require("react-redux"), "useDispatch")
      .mockReturnValue([...mockChatMessages]);

    const { container } = renderWithProviders(
      <Chat chattingWith={mockChattingWith} currentUser={mockCurrentUser} />
    );

    const usernameElement = screen.getByText(mockChattingWith.username);
    const message1Element = screen.getByText(mockChatMessages[0].message);
    const message2Element = screen.getByText(mockChatMessages[1].message);
    const message3Element = screen.getByText(mockChatMessages[2].message);
    const inputElement = screen.getByTestId(/type message here/i);
    const sendButtonElement = screen.getByRole("button");

    expect(usernameElement).toBeInTheDocument();

    await waitFor(() => {
      expect(message1Element).toBeInTheDocument();
      expect(message2Element).toBeInTheDocument();
      expect(message3Element).toBeInTheDocument();
    });

    expect(inputElement).toBeInTheDocument();
    expect(sendButtonElement).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("renders 'Say Hi' when there's no message history and calls sendMessage on 'Say Hi' click", async () => {
    jest.spyOn(require("react-redux"), "useSelector").mockReturnValue([]);
    jest.spyOn(require("react-redux"), "useDispatch").mockReturnValue([]);
    const sendMessageSpy = jest.spyOn(ApiMethods, "sendMessage");

    const { container } = renderWithProviders(
      <Chat chattingWith={mockChattingWith} currentUser={mockCurrentUser} />
    );

    const sayHiElement = screen.getByText(/say hi!/i);

    await waitFor(() => {
      expect(sayHiElement).toBeInTheDocument();
    });

    await userEvent.click(sayHiElement);

    await waitFor(() => {
      expect(sendMessageSpy).toBeCalledWith({
        content: "Hi!",
        sender_id: mockCurrentUser.id,
        receiver_id: mockChattingWith.id,
      });
    });

    expect(container).toMatchSnapshot();
  });

  test("calls sendMessage with message object on send button click", async () => {
    jest
      .spyOn(require("react-redux"), "useSelector")
      .mockReturnValue([...mockChatMessages]);
    jest
      .spyOn(require("react-redux"), "useDispatch")
      .mockReturnValue([...mockChatMessages]);
    const sendMessageSpy = jest.spyOn(ApiMethods, "sendMessage");

    renderWithProviders(
      <Chat chattingWith={mockChattingWith} currentUser={mockCurrentUser} />
    );

    const inputElement = screen.getByTestId(/type message here/i);
    const sendButtonElement = screen.getByRole("button");

    await userEvent.type(inputElement, "This is a test");
    await userEvent.click(sendButtonElement);

    await waitFor(() => {
      expect(sendMessageSpy).toBeCalledWith({
        content: "This is a test",
        sender_id: mockCurrentUser.id,
        receiver_id: mockChattingWith.id,
      });
    });
  });
});

describe("ActionCable", () => {
  test("invokes received method when new message received", async () => {
    let mockMessages = [...mockChatMessages];

    jest
      .spyOn(require("react-redux"), "useSelector")
      .mockImplementation((selector) =>
        selector({ chat: { messages: mockMessages } })
      );

    const mockDispatch = jest.fn();

    jest
      .spyOn(require("react-redux"), "useDispatch")
      .mockReturnValue(mockDispatch);

    const mockReceived = jest.fn();
    const mockUnsubscribe = jest.fn();

    const mockSubscription = {
      create: jest.fn().mockReturnValue({
        received: mockReceived,
        unsubscribe: mockUnsubscribe,
      }),
    };

    const createConsumerSpy = jest.spyOn(ActionCable, "createConsumer");

    createConsumerSpy.mockReturnValueOnce({
      subscriptions: {
        ...mockSubscription,
      },
    });

    renderWithProviders(
      <Chat chattingWith={mockChattingWith} currentUser={mockCurrentUser} />
    );

    await waitFor(() => {
      expect(createConsumerSpy).toBeCalled();
      expect(mockSubscription.create).toHaveBeenCalledWith(
        {
          channel: "ChatChannel",
          sender_id: mockCurrentUser.id,
          receiver_id: mockChattingWith.id,
        },
        {
          received: expect.any(Function),
        }
      );
    });

    act(() => {
      mockSubscription.create.mock.calls[0][1].received(mockNewMessage);
    });

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(
        setMessages([
          ...mockChatMessages,
          {
            id: mockNewMessage.id,
            message: mockNewMessage.content,
            position: "single",
            receiver: "User2",
            receiver_id: mockNewMessage.receiver_id,
            sender: "User1",
            sender_id: mockNewMessage.sender_id,
            direction: mockNewMessage.sender_id === 1 ? "outgoing" : "incoming",
          },
        ])
      );
    });
  });
});
