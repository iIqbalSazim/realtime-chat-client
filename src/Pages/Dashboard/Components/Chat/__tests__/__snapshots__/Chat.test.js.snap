// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders 'Say Hi' when there's no message history and calls sendMessage on 'Say Hi' click 1`] = `
<div>
  <div
    class="cs-chat-container"
  >
    <div
      class="cs-conversation-header mb-2"
    >
      <div
        class="cs-conversation-header__content"
      >
        <div
          class="cs-conversation-header__user-name"
        >
          User2
        </div>
        <div
          class="cs-conversation-header__info"
        />
      </div>
    </div>
    <div
      class="cs-message-list"
    >
      <div
        class="scrollbar-container cs-message-list__scroll-wrapper ps"
        data-cs-message-list=""
      >
        <div
          class="items-center flex justify-center w-full my-2"
        >
          <div
            class="btn btn-neutral btn-outline btn-wide"
          >
            Say Hi!
          </div>
        </div>
        <div
          class="cs-message-list__scroll-to"
        />
        <div
          class="ps__rail-x"
          style="left: 0px; top: 0px;"
        >
          <div
            class="ps__thumb-x"
            style="left: 0px; width: 0px;"
            tabindex="0"
          />
        </div>
        <div
          class="ps__rail-y"
          style="top: 0px; left: 0px;"
        >
          <div
            class="ps__thumb-y"
            style="top: 0px; height: 0px;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
    <div
      class="cs-message-input"
      data-testid="message-input-box"
    >
      <div
        class="cs-message-input__content-editor-wrapper"
      >
        <div
          class="scrollbar-container cs-message-input__content-editor-container ps"
        >
          <div
            class="cs-message-input__content-editor"
            contenteditable="true"
            data-placeholder="Type message here"
          />
          <div
            class="ps__rail-x"
            style="left: 0px; top: 0px;"
          >
            <div
              class="ps__thumb-x"
              style="left: 0px; width: 0px;"
              tabindex="0"
            />
          </div>
          <div
            class="ps__rail-y"
            style="top: 0px; left: 0px;"
          >
            <div
              class="ps__thumb-y"
              style="top: 0px; height: 0px;"
              tabindex="0"
            />
          </div>
        </div>
      </div>
      <div
        class="cs-message-input__tools"
      >
        <button
          class="cs-button cs-button--send"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane fa-w-16 "
            data-icon="paper-plane"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Chat renders correctly with populated messages 1`] = `
<div>
  <div
    class="cs-chat-container"
  >
    <div
      class="cs-conversation-header mb-2"
    >
      <div
        class="cs-conversation-header__content"
      >
        <div
          class="cs-conversation-header__user-name"
        >
          User2
        </div>
        <div
          class="cs-conversation-header__info"
        />
      </div>
    </div>
    <div
      class="cs-message-list"
    >
      <div
        class="scrollbar-container cs-message-list__scroll-wrapper ps"
        data-cs-message-list=""
      >
        <section
          class="cs-message cs-message--outgoing cs-message--single"
          data-cs-message=""
        >
          <div
            class="cs-message__content-wrapper"
          >
            <div
              class="cs-message__header"
            >
              <div
                class="cs-message__sender-name"
              />
              <div
                class="cs-message__sent-time"
              />
            </div>
            <div
              class="cs-message__content"
            >
              <div
                class="cs-message__html-content"
              >
                Hello
              </div>
            </div>
          </div>
        </section>
        <section
          class="cs-message cs-message--incoming cs-message--single"
          data-cs-message=""
        >
          <div
            class="cs-message__content-wrapper"
          >
            <div
              class="cs-message__header"
            >
              <div
                class="cs-message__sender-name"
              />
              <div
                class="cs-message__sent-time"
              />
            </div>
            <div
              class="cs-message__content"
            >
              <div
                class="cs-message__html-content"
              >
                Hey there
              </div>
            </div>
          </div>
        </section>
        <section
          class="cs-message cs-message--outgoing cs-message--single"
          data-cs-message=""
        >
          <div
            class="cs-message__content-wrapper"
          >
            <div
              class="cs-message__header"
            >
              <div
                class="cs-message__sender-name"
              />
              <div
                class="cs-message__sent-time"
              />
            </div>
            <div
              class="cs-message__content"
            >
              <div
                class="cs-message__html-content"
              >
                How are you?
              </div>
            </div>
          </div>
        </section>
        <div
          class="cs-message-list__scroll-to"
        />
        <div
          class="ps__rail-x"
          style="left: 0px; top: 0px;"
        >
          <div
            class="ps__thumb-x"
            style="left: 0px; width: 0px;"
            tabindex="0"
          />
        </div>
        <div
          class="ps__rail-y"
          style="top: 0px; left: 0px;"
        >
          <div
            class="ps__thumb-y"
            style="top: 0px; height: 0px;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
    <div
      class="cs-message-input"
      data-testid="message-input-box"
    >
      <div
        class="cs-message-input__content-editor-wrapper"
      >
        <div
          class="scrollbar-container cs-message-input__content-editor-container ps"
        >
          <div
            class="cs-message-input__content-editor"
            contenteditable="true"
            data-placeholder="Type message here"
          />
          <div
            class="ps__rail-x"
            style="left: 0px; top: 0px;"
          >
            <div
              class="ps__thumb-x"
              style="left: 0px; width: 0px;"
              tabindex="0"
            />
          </div>
          <div
            class="ps__rail-y"
            style="top: 0px; left: 0px;"
          >
            <div
              class="ps__thumb-y"
              style="top: 0px; height: 0px;"
              tabindex="0"
            />
          </div>
        </div>
      </div>
      <div
        class="cs-message-input__tools"
      >
        <button
          class="cs-button cs-button--send"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane fa-w-16 "
            data-icon="paper-plane"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
